Support filtering top-level resource based on embedded resource filter

steve-chavez opened the pull request:
Closes #1075. Allow filtering the top-level resource based on the embedded filter ** ~~with the `LATERAL/GROUP BY` approach mentioned on https://github.com/PostgREST/postgrest/issues/1075#issuecomment-381716276,~~ like so:

```bash
GET /projects?select=id,clients!inner(id)&clients.id=eq.2

[{"id":3,"client":{"id":2}},
 {"id":4,"client":{"id":2}}]

# the usual left join would give
GET /projects?select=id,clients(id)&clients.id=eq.2
[{"id":1,"clients":null},
{"id":2,"clients":null},
{"id":3,"clients":{"id":2}},
{"id":4,"clients":{"id":2}},
{"id":5,"clients":null}]
```

~~This PR will **only cover tables**, views will need more work and smartness from our schema cache; mainly because when doing `GROUP BY`, pg is able to infer the pk on a table but not a view, more details [here](https://dba.stackexchange.com/questions/195104/postgres-group-by-id-works-on-table-directly-but-not-on-identical-view).~~

**Edit**: Thanks to Iced-Sun's query below(https://github.com/PostgREST/postgrest/pull/1949#issuecomment-924191966), this will now work for both tables and views.

This behavior can also  be enabled by default with the following config option

```
db-embed-default-join="inner"
```

Which saves the need for specifying `!inner` on every request. In this case, you can go back to the previous behavior per request by specifying `!left`  on the embedded resource, e.g `/projects?select=*,clients!left(*)&clients.id=eq.12`


## Steps

- [x] many-to-one relationships
- [x] one-to-many relationships
- [x] many-to-many relationships
- [x] decide on syntax for the inner join
- [x] config for default embed mode
- [x] CHANGELOG

---

This PR won't cause a breaking change. Embedding will still use subqueries by default, only on `!inner` it will use LATERAL.

wolfgangwalther commented:
Sorry for not bringing this up earlier in the issue, but looking at the example ...

```http
GET /projects?select=id,clients!inner(id)&clients.id=eq.2
```

... it suddenly strikes me as odd to have something in the `select=` part change the number of rows. Imho, the `select=` part should only change the shape of the output / each row - but should not filter anything.

Since we already have embedded filters as `<alias>.<column>=` - how about extending that to support a special operator on only `<alias>=`?

Maybe something like this could make it an inner join:

```http
GET /projects?select=id,clients(id)&clients.id=eq.2&clients=inner
```

steve-chavez commented:
> ... it suddenly strikes me as odd to have something in the select= part change the number of rows. Imho, the select= part should only change the shape of the output / each row - but should not filter anything.

@wolfgangwalther True. I was mostly going from what we discussed in https://github.com/PostgREST/postgrest/issues/915#issuecomment-780035727. Mainly that the embedding was a function and that the syntax for parameters would be `func!param1!param2` so `embed!join_type!hint(..)`.

> Since we already have embedded filters as <alias>.<column>= - how about extending that to support a special operator on only <alias>=?

I like the direction. Maybe we could have have something like:

```http
GET /projects?select=id,clients(id)&clients.id=eq.2&clients=join.inner
```

In fact, the hints could also be expressed on a filter:

```http
GET /projects?select=id,clients(id)&clients.id=eq.2&clients=join(inner).hint1,hint2
```

(The `join(inner)`) syntax would be similar to our [full text search](https://postgrest.org/en/v8.0/api.html#full-text-search) with the lang)

That keeps the `select` param cleaner. It could also provide a path for doing the future function syntax:

```http
GET ?select=avg&avg=args.amount
```

steve-chavez commented:
> GET /projects?select=id,clients(id)&clients.id=eq.2&clients=join(inner).hint1,hint2

I guess one drawback of that syntax is that it's more verbose(both for the REST interface and libraries probably).

wolfgangwalther commented:
> Mainly that the embedding was a function and that the syntax for parameters would be `func!param1!param2` so `embed!join_type!hint(..)`.

I see. The "embedding as a function" pov only holds true for the `SELECT` part, though - at least if we think of "function" as a "PostgreSQL `FUNCTION`". This is because we can't replicate an inner join with a computed column function.

> I like the direction. Maybe we could have have something like:
> 
> ```httpspec
> GET /projects?select=id,clients(id)&clients.id=eq.2&clients=join.inner
> ```

Hm, I think it's enough to specify the `inner` only. That's assuming `join` is the only operator anyway. I can imagine we could support `inner` and `anti` (for an ANTI-JOIN) down the road. Not sure about whether `right` and/or `full` make any sense. But I don't see anything other than `join`, so we might as well omit it.


> In fact, the hints could also be expressed on the query param:
> 
> ```httpspec
> GET /projects?select=id,clients(id)&clients.id=eq.2&clients=join(inner).hint1,hint2
> ```
> 
> (The `join(inner)`) syntax would be similar to our [full text search](https://postgrest.org/en/v8.0/api.html#full-text-search) with the lang)
> 
> That keeps the `select` param cleaner. It could also provide a path for doing the future function syntax:
> 
> ```httpspec
> GET ?select=avg&avg=args.amount
> ```

Assuming we'd keep the ability to put hints (and therefore the embedding specification) in only one place (either query param or select), this would be a breaking change.

Once we're at that stage, where we implement such a big breaking change to the query syntax, I'd have a few other points to raise to change in the query syntax. Two biggest things that come to mind are correct quoting / escaping and moving the operator to the left of the equal sign like e.g. `column.eq=value`. This would make things a fair bit easier for plain-html-forms, to support no javascript environments. Also the column+operator used tends to be quite static, while the value used for the query is not - separating the static part from the dynamic part via `=` (which is supported by all the client librarys for query params) seems the logical choice.

So, I suggest to just go with the non-breaking change for this PR - and then open another issue where we can discuss a "new query syntax" that could get a few things straight. We could implement both at the same time and have them configurable via config option, to be able to slowly deprecate the old syntax.

steve-chavez commented:
> Hm, I think it's enough to specify the inner only. That's assuming join is the only operator anyway. I can imagine we could support inner and anti (for an ANTI-JOIN) down the road. But I don't see anything other than join, so we might as well omit it.

It would be too inconsistent at the parsers level, I'm not even sure if it can be done cleanly. All of our filters have a `prefix.`, with the sole exception of function args, which where originally meant to be called with the `arg.` prefix(`arg1=arg.value`) but that was decided against because of verbosity. This is done by checking if the rpc route is being used(all filters without prefix are args).

> and moving the operator to the left of the equal sign like e.g. column.eq=value. This would make things a fair bit easier for plain-html-forms, to support no javascript environments.

Hm, if this is meant for the `eq` filter, we could have a similar rule as the one mentioned above regarding the function arg: for table routes, default to `eq` for filters with no `prefix.`(so `id=1` would work). This has been asked before on an issue IIRC, also on [reddit](https://www.reddit.com/r/programming/comments/jymp14/postgrest_rest_api_for_any_postgres_database/ggn1p8f/?utm_source=reddit&utm_medium=web2x&context=3). There would be no big breaking change for this.

> Assuming we'd keep the ability to put hints (and therefore the embedding specification) in only one place (either query param or select), this would be a breaking change.

I was thinking to have both actually, so no breaking change. I'm still not set on either though. For now I'll continue with the `select=..!inner`, doing the queries is the hardest part about this PR.

wolfgangwalther commented:
> Hm, if this is meant for the `eq` filter,

No, I propose to have all filters on the left side. `id.lt=5`, `id.gt=5`, ...

> I was thinking to have both actually, so no breaking change.

Imho, that would make it quite complicated to parse. You could have conflicting targets/hints/... in both parts - this will be very hard to efficiently create a good query.

Iced-Sun commented:
The `LATERAL/GROUP BY` approach for resource embedding with inner join depends on the fact that pg could group by pk even when the group keys include other columns (`group by pk, c1, c2, c3` => `group by pk`). But the inference doesn't happen for a view, or a `not null unique` key, and hence a potentially expensive full group-by must be applied.

I propose to implement resource embedding with correlated join, which is essentially the same as the current approach correlated subquery, but with the extensibility to support left/inner/anti-joins.

Specifically, instead of translating the left-join resource embedding of `GET /projects?select=id,clients(id)&clients.id=eq.2` to
```sql
with pg_source as (
  select projects.id, coalesce(
           (select json_agg(clients.*)
              from (select clients.id 
                      from clients
                     where clients.project_id = projects.id and clients.id = 2
              ) clients), 
         '[]') as clients
    from projects
)
select coalesce(json_agg(_postgrest_t), '[]')::character varying as body
from (select * from pg_source) _postgrest_t;
```

a correlated join version could be
```sql
with pg_source as (
     select projects.id, coalesce(nullif(clients_clients.clients::text, '[null]'), '[]')::json AS clients
       from projects
  left join lateral (select json_agg(clients) clients 
                       from (select clients.id 
                               from clients
                              where clients.project_id = projects.id and clients.id = 2
                            ) clients
                    ) clients_clients
         on clients_clients.clients is not null
)
select coalesce(json_agg(_postgrest_t), '[]')::character varying as body
from (select * from pg_source) _postgrest_t;
```

An inner-join resource embedding is then 
```sql
with pg_source as (
     select projects.id, coalesce(nullif(clients_clients.clients::text, '[null]'), '[]')::json AS clients
       from projects
       join lateral (select json_agg(clients) clients 
                       from (select clients.id 
                               from clients
                              where clients.project_id = projects.id and clients.id = 2
                            ) clients
                    ) clients_clients
         on clients_clients.clients is not null
)
select coalesce(json_agg(_postgrest_t), '[]')::character varying as body
from (select * from pg_source) _postgrest_t;
```

And an anti-join resource embedding is 
```sql
with pg_source as (
     select projects.id, coalesce(nullif(clients_clients.clients::text, '[null]'), '[]')::json AS clients
       from projects
       join lateral (select json_agg(clients) clients 
                       from (select clients.id 
                               from clients
                              where clients.project_id = projects.id and clients.id = 2
                            ) clients
                    ) clients_clients
         on clients_clients.clients is null
)
select coalesce(json_agg(_postgrest_t), '[]')::character varying as body
from (select * from pg_source) _postgrest_t;
```

The correlated-join approach doesn't rely on pk inference, works on view and `not null unique` key. And the performance is essentially the same as subquery (https://github.com/PostgREST/postgrest/issues/1075#issuecomment-569489136).

Does that make any sense?

Iced-Sun commented:
BTW, the ability to do aggregation and flatten join seems to be a popular requested feature (#1233, #211, #1126, #915 and others). Do we have any plan to support explicit join/group by?

Say, something like 
1. `GET /projects,projects_traits?select=projects.id,projects_traits.category`
2. `GET /projects?select=count(*)&groupby=zone`
3. `GET /projects,project_traits?select=count(*),projects_traits.category&groupby=projects_traits.category`

Look like it will take a huge effort.

steve-chavez commented:
@Iced-Sun Awesome! Your approach is much better since it doesn't rely on GROUP BY and it also works on views :tada: :tada: 

So to understand it better, taking my approach on https://github.com/PostgREST/postgrest/issues/1075#issuecomment-381716276, the query for inner joins would now be

```sql
WITH pg_source AS (
   SELECT "test"."clients"."id",
          "test"."clients"."name",
          "projects_projects" AS "projects"
   FROM "test"."clients"
   INNER JOIN LATERAL (
     SELECT json_agg("projects") "projects" -- important change here
     FROM(
       SELECT "test"."projects"."id", "test"."projects"."name"
       FROM "test"."projects"
       WHERE "test"."projects"."client_id" = "test"."clients"."id") "projects"
   ) AS "projects_projects" ON "projects_projects"."projects" IS NOT NULL -- important change here
)
SELECT coalesce(json_agg(_postgrest_t), '[]')::character varying AS BODY
FROM (SELECT * FROM pg_source) _postgrest_t;
```

Where the main changes are the correlated join(which avoids the need for GROUP BY) plus the `LATERAL ... ON <target> IS NOT NULL`.

As you pointed out, this practically has the same cost that our correlated subquery has(`cost=2229.33..2229.34`). The correlated join cost(`2261.08..2261.09`) is higher than the GROUP BY(`125.21..125.22`) query but of course it doesn't have any of the downsides - which is definitely good enough.

(The above is just my simplistic view, I think you've determined that GROUP BY is actually worse in perf)

So I'll proceed with the correlated join approach for this feature.

> Do we have any plan to support explicit join/group by?

Yes, but besides the syntax, we still need a way to protect against a potentially expensive group by, this was discussed on https://github.com/PostgREST/postgrest/issues/915.

Iced-Sun commented:
> As you pointed out, this practically has the same cost that our correlated subquery has(`cost=2229.33..2229.34`). The correlated join cost(`2261.08..2261.09`) is higher than the GROUP BY(`125.21..125.22`) query but of course it doesn't have any of the downsides - which is definitely good enough.
> 
> (The above is just my simplistic view, I think you've determined that GROUP BY is actually worse in perf)

I did some simple performance tests in https://github.com/PostgREST/postgrest/issues/1075#issuecomment-569489136. It suggests that the `GROUP BY` approach is much more performent in the absence of a proper index; while with `create index on project (client_id)`, the correlated join (and correlated subquery) overtakes marginally (but not so marginal for a small result set). So I think `group by` may be preferable if we didn't have the pk problem to consider because it is more stable (on perf).

I have no solid facts on grouping by full-size keys (in the case that group-by-pk is not applicable), but I do have some not-so-good experiences when I have to group by multiple text columns to work around bad schema design.

As you said,  correlated join doesn't break things or bring surprises, and it unifies the left/inner/anti-join resource embedding.

Thanks for the great piece of software.

steve-chavez commented:
> .. it suddenly strikes me as odd to have something in the select= part change the number of rows. Imho, the select= part should only change the shape of the output / each row - but should not filter anything.

@wolfgangwalther Been thinking about this, why is it "odd" to change the shape of the output on `select`? Because when using count(and we'd have more aggregates later) this happens:

```http
GET /projects?select=count

[{"count":5}]
```

And of course in SQL the same thing happens - not only WHERE modifies the rows, SELECT also does.

wolfgangwalther commented:
> Been thinking about this, why is it "odd" to change the shape of the output on `select`?
> [...]
> [...] in SQL the same thing happens - not only WHERE modifies the rows, SELECT also does.

That's odd for me, already. For me, using an aggregate function without a `GROUP BY` basically adds an implicit `GROUP BY` with a constant value. I would have liked that to be more explicit in SQL, too. But that ship has sailed for a long time, I guess.

steve-chavez commented:
> That's odd for me, already. For me, using an aggregate function without a GROUP BY basically adds an implicit GROUP BY with a constant value

@wolfgangwalther I see. I've also always find it odd that these queries work

```sql
select p.json_agg from test.projects p;
select p.count from projects p;
select p.array_agg from projects p;
-- same through postgrest
```
(I think we discussed this somewhere before)

> I would have liked that to be more explicit in SQL, too. But that ship has sailed for a long time, I guess.

Yeah. So considering this, WDYT about just going with the `select=*,embed!inner(*)` syntax? Considering `select` is not consistent, having an `embed=join.inner` syntax only adds more verbosity.

steve-chavez commented:
Alright, this is now ready for a final review. I've added tests for views(with m2o/o2m/m2m cases), rpc and embedding with hints.

I've also added a `db-embed-default-join` config option(mentioned above https://github.com/PostgREST/postgrest/pull/1949#issue-997379608) for users that want this join type by default.

wolfgangwalther commented on CHANGELOG.md:
```diff
@@ -14,6 +14,10 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 - #1938, Allow escaping inside double quotes with a backslash, e.g. `?col=in.("Double\"Quote")`, `?col=in.("Back\\slash")` - @steve-chavez
+- #1075, Allow filtering top-level resource based on embedded resources filters - @steve-chavez, @Iced-Sun
+   + This is enabled by adding `!inner` to the embedded resource, e.g. `/projects?select=*,clients!inner(*)&clients.id=eq.12`
+   + This behaviour can be enabled by default with the `db-embed-default-join='inner'` config option, which saves the need for specifying `!inner` on every request.
+     In this case, you can go back to the previous behavior per request by specifying `!left`  on the embedded resource, e.g `/projects?select=*,clients!left(*)&clients.id=eq.12`
```
```suggestion
   + This behaviour can be enabled by default with the `db-embed-default-join='inner'` config option, which saves the need for specifying `!inner` on every request. In this case, you can go back to the previous behavior per request by specifying `!left` on the embedded resource, e.g `/projects?select=*,clients!left(*)&clients.id=eq.12`
```

steve-chavez commented on CHANGELOG.md:
```diff
@@ -579,4 +583,4 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 ### Fixed
 - Make filter position match docs, e.g. `?order=col.asc` rather
-  than `?order=asc.col`.
+  than `?order=asc.col`.
```
Yeah, can I get a pass on this one :smiley: ? I've configured vim to strip unnecessary whitespace each time I save a file. If I bring back the whitespace I'm going to have the same problem each time I edit the CHANGELOG :unamused: .

colearendt commented:
How does this work with joins that are already using `!` for the foreign key in use? Is it possible to specify a foreign key _and_ an inner join?

i.e. if I have this:
```
select=*,alias:clients!fk_main_clients(*)
```

How should it switch to an `inner` join?

colearendt commented:
Thanks! That's perfect! I totally missed that example 🙈  Found a little bug in #1977, but this is really awesome!! Well done!
